apps/web/server/api/routers/device-auth.ts

```diff
@@ -0,0 +1,444 @@
+ *
+ * Flow:
+ * 1. CLI calls `initiate` to get device_code + user_code
+ * 2. User visits /device page and enters user_code
+ * 3. User calls `verify` to link the device code to their session
+ * 4. CLI polls `poll` until verification completes, then receives session token
+ */
+export const deviceAuthRouter = createTRPCRouter({
+  /**
+   * Initiate device auth flow (called by CLI)
+   *
+   * Generates a device code (UUID) and user code (8-char alphanumeric).
+   * Returns absolute verification URLs per RFC 8628.
+   */
+  initiate: publicProcedure.mutation(async ({ ctx }) => {
+    const deviceCode = crypto.randomUUID();
+    const userCode = generateUserCode(8);
+    const expiresAt = new Date(Date.now() + CODE_EXPIRY_MINUTES * 60 * 1000);
+
+    // Insert the device auth code
+    await ctx.db.insert(schema.deviceAuthCodes).values({
+      deviceCode,
+      userCode,
+      expiresAt,
+      isUsed: false,
+    });
+
+    // Format user code with dash for display (XXXX-XXXX)
+    const formattedUserCode = `${userCode.slice(0, 4)}-${userCode.slice(4)}`;
+
+    // Construct absolute URLs per RFC 8628
+    const baseUrl = getVerificationBaseUrl(ctx.headers);
+    const verificationUri = `${baseUrl}${DEVICE_VERIFICATION_PATH}`;
+
+    // Use URL API for proper encoding
+    const completeUrl = new URL(verificationUri);
+    completeUrl.searchParams.set("user_code", userCode);
+    const verificationUriComplete = completeUrl.toString();
+
+    return {
+      deviceCode,
+      userCode: formattedUserCode,
+      verificationUri,
+      verificationUriComplete,
+      expiresIn: CODE_EXPIRY_MINUTES * 60,
+      interval: POLL_INTERVAL_SECONDS,
+    };
+  }),
+
+  /**
+   * Verify a device code (called by browser after user enters code)
+   *
+   * This is a protected procedure - user must be logged in.
+   * Links the device code to the authenticated user and creates a CLI session.
+   */
+  verify: protectedProcedure
+    .input(
+      z.object({
+        userCode: z
+          .string()
+          .min(1, "User code is required")
+          .transform((val) => val.replace(/-/g, "").trim()),
+      }),
+    )
+    .mutation(async ({ ctx, input }) => {
+      const now = new Date();
+
+      // Generate high-entropy session token (256 bits)
+      // SECURITY: This token is a credential - never log it
+      const cliSessionToken = generateCliSessionToken();
+      const sessionExpiresAt = new Date(
+        Date.now() + SESSION_EXPIRY_DAYS * 24 * 60 * 60 * 1000,
+      );
+
+      // Try to get browser session ID for reference (optional)
+      // This links the CLI session to the browser session that authorized it
+      const browserSessionId = ctx.session?.user
+        ? await getBrowserSessionId(ctx.db, ctx.user.id)
+        : null;
+
+      // Create CLI session first (required by FK constraint on device_auth_codes)
+      await ctx.db.insert(schema.cliSessions).values({
+        id: cliSessionToken,
```

Comment:
Yeah I think in general any place where we can let the database generate these codes, the better